packages/components/addon/modifiers/hds-clipboard.js

```diff
@@ -0,0 +1,177 @@
+      );
+      return;
+    }
+  } else if (target instanceof Node && target.nodeType === 1) {
+    targetElement = target;
+  } else {
+    if (target instanceof NodeList) {
+      assert(
+        '`hds-clipboard` modifier - `target` argument must be a string or a DOM node - provided: a list of DOM nodes'
+      );
+    } else {
+      assert(
+        `\`hds-clipboard\` modifier - \`target\` argument must be a string or a DOM node - provided: ${typeof target}`
+      );
+    }
+  }
+  return targetElement;
+};
+
+export const getTextToCopyFromTargetElement = (targetElement) => {
+  let textToCopy;
+  if (targetElement instanceof Node && targetElement.nodeType === 1) {
+    if (
+      targetElement instanceof HTMLInputElement || // targetElement.nodeName === 'INPUT' ||
+      targetElement instanceof HTMLTextAreaElement || // targetElement.nodeName === 'TEXTAREA' ||
+      targetElement instanceof HTMLSelectElement // targetElement.nodeName === 'SELECT'
+    ) {
+      textToCopy = targetElement.value;
+    } else {
+      // simplest approach
+      textToCopy = targetElement.innerText;
+
+      // approach based on text selection (left for backup just in case)
+      // var selection = window.getSelection();
+      // var range = document.createRange();
+      // selection.removeAllRanges();
+      // range.selectNodeContents(targetElement);
+      // selection.addRange(range);
+      // textToCopy = selection.toString();
+      // selection.removeAllRanges();
+    }
+  }
+  return textToCopy;
+};
+
+export const writeTextToClipboard = async (textToCopy) => {
+  // finally copy the text to the clipboard using the Clipboard API
+  // https://developer.mozilla.org/en-US/docs/Web/API/Clipboard_API
+  if (textToCopy) {
+    try {
+      // notice: the "clipboard-write" permission is granted automatically to pages when they are in the active tab
+      // https://developer.mozilla.org/en-US/docs/Web/API/Clipboard/write
+      await navigator.clipboard.writeText(textToCopy);
+      // DEBUG uncomment this for easy debugging
+      // console.log('success', textToCopy);
+      return true;
+    } catch (error) {
+      // clipboard write failed
+      console.log(
+        '`hds-clipboard` modifier - an error occurred while writing to the clipboard - check your browser permissions',
+        textToCopy,
+        error
+      );
+      return false;
+    }
+  } else {
+    return false;
+  }
+};
+
+export const copyToClipboard = async (text, target) => {
+  let textToCopy;
+
+  if (text) {
+    textToCopy = getTextToCopy(text);
+  } else if (target) {
+    const targetElement = getTargetElement(target);
+    textToCopy = getTextToCopyFromTargetElement(targetElement);
+  } else {
+    assert(
+      '`hds-clipboard` modifier - either a `text` or a `target` argument is required'
+    );
+  }
+  const success = await writeTextToClipboard(textToCopy);
+  return success;
+};
+
+export default class HdsClipboardModifier extends Modifier {
```

Comment:
I think this whole thing could be a function modifier, you'd create the handler as a local var inline to close over the args, but that would also let you have a reference to it for cleanup in a function you return.

Reply:
From discussion with @meirish (for future reference)
- MI: I don’t know why it was a class-based modifier to start
- CR: Because I used this as model: https://github.com/hashicorp/design-system/blob/main/packages/components/addon/modifiers/hds-tooltip.js
Which was based off this: https://github.com/hashicorp/cloud-ui/blob/main/addons/core/addon/modifiers/tooltip.js
- MI: https://github.com/ember-modifier/ember-modifier#function-based-modifiers has some documentation which can show the pros/cons of function-based vs. class-based. (class-based provides more control and allows you to inject a service - but you don't need either of those here imo

Thanks for the feedback/clarification Matthew! 🙏